satpy/resample.py

```diff
@@ -823,11 +823,187 @@ def compute(self, data, expand=True, **kwargs):
                             coords=coords or None)
 
 
+class BucketResamplerBase(BaseResampler):
+    """Base class for bucket resampling which implements averaging.
+    """
+
+    def __init__(self, source_geo_def, target_geo_def):
+        super(BucketResamplerBase, self).__init__(source_geo_def, target_geo_def)
+        self.resampler = None
+
+    def precompute(self, **kwargs):
+        """Create X and Y indices and store them for later use."""
+        LOG.debug("Initializing bucket resampler.")
+        source_lons, source_lats = self.source_geo_def.get_lonlats(
+            chunks=CHUNK_SIZE)
+        self.resampler = bucket.BucketResampler(self.target_geo_def,
+                                                source_lons,
+                                                source_lats)
+
+    def compute(self, data, **kwargs):
+        """Call the resampling."""
+        raise NotImplementedError("Use the sub-classes")
+
+    def resample(self, data, **kwargs):
+        """Resample `data` by calling `precompute` and `compute` methods.
+
+        Args:
+            data (xarray.DataArray): Data to be resampled
+
+        Returns (xarray.DataArray): Data resampled to the target area
+        """
+        self.precompute(**kwargs)
+        attrs = data.attrs.copy()
+        data_arr = data.data
+        if data.ndim == 3 and data.dims[0] == 'bands':
+            dims = ('bands', 'y', 'x')
+        # Both one and two dimensional input data results in 2D output
+        elif data.ndim in (1, 2):
+            dims = ('y', 'x')
+        else:
+            dims = data.dims
+        result = self.compute(data_arr, **kwargs)
+        coords = {}
+        if 'bands' in data.coords:
+            coords['bands'] = data.coords['bands']
+        # Fractions are returned in a dict
+        elif isinstance(result, dict):
+            coords['categories'] = sorted(result.keys())
+            dims = ('categories', 'y', 'x')
+            new_result = []
+            for cat in coords['categories']:
+                new_result.append(result[cat])
+            result = da.stack(new_result)
+        if result.ndim > len(dims):
+            result = da.squeeze(result)
+
+        # Adjust some attributes
+        if "BucketFraction" in str(self):
+            attrs['units'] = None
+            attrs['calibration'] = None
+            attrs['standard_name'] = 'area_fraction'
+        elif "BucketCount" in str(self):
+            attrs['units'] = None
+            attrs['calibration'] = None
+            attrs['standard_name'] = 'number_of_observations'
```

Comment:
Are there actual CF standard names for these types of data?

Reply:
The `'number_of_observations'` and `'area_fraction'` (on the previous lines not shown above quote) are from CF. For `'units'` and `'calibration'` I couldn't find anything that stands for "no relevant value". Is `None` ok, or should it be an empty string `''`?

Reply:
unitless in CF is either `'1'` or `''`. I suppose we could use `''` for calibration too.

Reply:
Thanks. I changed `None` to `''` for both attributes.

Reply:
Do we want to be consistent across Satpy and use `'1'` everywhere for unitless fields?

Reply:
`'1'` makes sense for numerical fields, not so much for strings...

Reply:
Good enough for me.